### ISSUE TITLE
[wontfix] minio hdfs gateway does not support uploading multiparts in parallel

### ISSUE THREAD

**Issue by mergewreck** [minio/minio#10012]
minio hdfs gateway does not support the multithreaded uploading of multiparts while minio server does.

The hdfs gateway raises exceptions like below:

_API: PutObjectPart(bucket=test, object=...)
Time: 12:50:18 CDT 07/10/2020
DeploymentID: 8e4a063a-d08d-4f5e-b283-63207880e299
RequestID: 1620757638A97928
RemoteHost: 127.0.0.1
Host: localhost:9000
UserAgent: MinIO (x86_64; x86_64) minio-java/dev
Error: append /.minio.sys/tmp/172175be-cbe9-46d4-9562-1c8e0426d5c6: append call failed with ERROR_APPLICATION (org.apache.hadoop.hdfs.protocol.AlreadyBeingCreatedException)
       4: github.com/minio/minio@/cmd/gateway/hdfs/gateway-hdfs.go:266:hdfs.hdfsToObjectErr()
       3: github.com/minio/minio@/cmd/gateway/hdfs/gateway-hdfs.go:684:hdfs.(*hdfsObjects).PutObjectPart()
       2: github.com/minio/minio@/cmd/object-handlers.go:2196:cmd.objectAPIHandlers.PutObjectPartHandler()
       1: net/http/server.go:2007:http.HandlerFunc.ServeHTTP()_

**Comment by harshavardhana** [comment]
Multipart is not supported on purpose HDFS gateway is designed for repatriation of data from HDFS to MinIO its scope is not to perpetually continue to use HDFS 

It has been added as part of roadmap https://github.com/minio/minio/blob/release/docs/gateway/hdfs.md#roadmap but we have decided to never do it.

**Comment by harshavardhana** [comment]
> Multipart is not supported on purpose HDFS gateway is designed for repatriation of data from HDFS to MinIO its scope is not to perpetually continue to use HDFS

If the requirement is always to use HDFS then use HDFS directly, there is no reason to add an S3 layer on top. The requirement for implementing HDFS gateway for us was to allow simple `mc mirror myhdfs/ myminio/` migration not the other way around.

**Comment by mergewreck** [comment]
Why not update the documentation:

[https://docs.min.io/docs/minio-gateway-for-hdfs.html](https://docs.min.io/docs/minio-gateway-for-hdfs.html)

**Known limitations**
Gateway inherits the following limitations of HDFS storage layer:

No bucket policy support (HDFS has no such concept)
No bucket notification APIs are not supported (HDFS has no support for fsnotify)
No server side encryption support (Intentionally not implemented)
No server side compression support (Intentionally not implemented)

**Roadmap**
Additional metadata support for PutObject operations
Additional metadata support for Multipart operations
Background append to provide concurrency support for multipart operations


A standard s3 api which can support s3 and hdfs (via gateway) is of great value in supporting cases where both are used- especially a case in which one is moving to s3 (and s3 apis) while moving away from hdfs. mc mirror only satisfies migration at the data laye, not the api layer.

**Comment by harshavardhana** [comment]
> A standard s3 api which can support s3 and hdfs (via gateway) is of great value in supporting cases where both are used- especially a case in which one is moving to s3 (and s3 apis) while moving away from hdfs. mc mirror only satisfies migration at the data laye, not the api layer.

Of course, I will update the docs right away to disallow misuse @mergewreck - That is not the scope of this gateway. If you want it that way you are free to fork and fix it yourself. Perpetually using HDFS keeping a hybrid approach is not our goal here.
